pkg/server/home_workspaces.go

```diff
@@ -0,0 +1,656 @@
+		}
+		result = result.Join(string(bucketBytes))
+	}
+
+	userName = reHomeWorkspaceNameDisallowedChars.ReplaceAllLiteralString(userName, "-")
+	userName = strings.TrimLeftFunc(userName, func(r rune) bool {
+		return r <= '9'
+	})
+	userName = strings.TrimRightFunc(userName, func(r rune) bool {
+		return r == '-'
+	})
+
+	return result.Join(userName)
+}
+
+// needsAutomaticCreation deduces, from the logical cluster name,
+// according the the expected home root and home bucket level number,
+// whether the corresponding workspace has to be checked for automatic creation
+// and what its workspace type will be.
+func (h *homeWorkspaceHandler) needsAutomaticCreation(logicalClusterName logicalcluster.Name) (needsCheck bool, workspaceType tenancyv1alpha1.ClusterWorkspaceTypeName) {
+	if !logicalClusterName.HasPrefix(h.homePrefix) {
+		return false, ""
+	}
+
+	if logicalClusterName == h.homePrefix {
+		return false, ""
+	}
+
+	levelsToHomePrefix := 0
+	for lcluster := logicalClusterName; lcluster != h.homePrefix; lcluster, _ = lcluster.Split() {
+		levelsToHomePrefix++
+	}
+
+	switch {
+	case levelsToHomePrefix <= h.bucketLevels:
+		return true, "Homebucket"
+	case levelsToHomePrefix == h.bucketLevels+1:
+		return true, "Home"
+	default:
+		return false, ""
+	}
+}
+
+// searchForReadyWorkspaceInLocalInformers checks whether a workspace is known on the current shard
+// in local informers.
+// For home workspaces, we also check:
+//   - if related RBAC resources are there,
+//   - if the workspace phase is READY
+// and if not answer to retry later.
+func searchForReadyWorkspaceInLocalInformers(h *homeWorkspaceHandler, logicalClusterName logicalcluster.Name, isHome bool, userName string) (found bool, retryAfterSeconds int, err error) {
+	workspace, err := h.localInformers.getClusterWorkspace(logicalClusterName)
+	if err != nil && !kerrors.IsNotFound(err) {
+		return false, 0, err
+	}
+	if workspace == nil {
+		return false, 0, nil
+	}
+
+	// Workspace has been found in local informer: check its status.
+	if workspaceUnschedulable(workspace) {
+		// The workspace cannot be scheduled => return the error.
+		return false, 0, kerrors.NewForbidden(tenancyv1alpha1.SchemeGroupVersion.WithResource("clusterworkspaces").GroupResource(), workspace.Name, errors.New("unschedulable workspace cannot be accessed"))
+	}
+
+	if !isHome {
+		// Waiting for home buckets to be ready is done in the `tryToCreate` function.
+		return true, 0, nil
+	}
+
+	if workspace.Status.Phase != tenancyv1alpha1.ClusterWorkspacePhaseReady {
+		// We have to wait for the workspace to be Ready before allowing actions in it,
+		// but only for the the home workspaces.
+		// Waiting for home buckets to be ready is done in the `tryToCreate` function.
+
+		// TODO (david): For now we accept all requests in this workspace when it's initializing...
+		// TODO (david): In the future, we might have a way to identify requests that are done through the
+		// initializer virtual workspace
```

Comment:
the users have the right to look into their initializing workspaces, and fix initialization if necessary. So we have to allow access. The initializing VW is just another access point for non-owners of the workspace.